apps/src/templates/progress/BubbleBadge.jsx

```diff
@@ -10,44 +10,76 @@ import FontAwesome from '../FontAwesome';
 const badgeTypes = ['assessment', 'keepWorking'];
 
 function BubbleBadge({type, isDiamond}) {
-  let backgroundColor, icon;
+  const position = isDiamond
+    ? styles.diamondBubblePosition
+    : styles.bubblePosition;
 
   if (type === 'assessment') {
-    backgroundColor = color.purple;
-    icon = 'check';
+    return <AssessmentBadge style={position} />;
   } else {
-    backgroundColor = color.red;
-    icon = 'exclamation';
+    return <KeepWorkingBadge showBorder={true} style={position} />;
   }
+}
 
-  const position = isDiamond
-    ? styles.diamondBubblePosition
-    : styles.bubblePosition;
+BubbleBadge.propTypes = {
+  type: PropTypes.oneOf(badgeTypes).isRequired,
+  isDiamond: PropTypes.bool,
+  positionedRelative: PropTypes.bool
+};
 
+export function KeepWorkingBadge({showBorder, style}) {
```

Comment:
this decomposition feels upside-down to me. we have `BubbleBadge` which chooses whether to render `KeepWorkingBadge` or `AssessmentBadge`, but then those two components have almost entirely duplicated code.

imo it would make more sense to put all the duplicated code in `BubbleBadge` (an internal component), then export both the `Assessment` and `KeepWorking` components, which are just responsible for configuring `BubbleBadge`. or at the very least move all this duplicated code into a new internal `BaseBadge` component.

Reply:
So the problem is that `BubbleBadge` also handles positioning for badge present on bubbles, which we don't want when we're using the `KeepWorkingBadge` outside of the context of a bubble. I'll do part of what you recommend and create a `BaseBadge`, I'll then export an `AssessmentBadge` and a `KeepWorkingBadge`, but I'll also export a `BubbleBadgeWrapper` which will handle the positioning of the badge for bubbles. On a bubble it will look like
```
<BubbleBadgeWrapper isDiamond={true}>
  <AssessmentBadge />
</BubbleBadgeWrapper>
```
Does this seem like a reasonable pattern?

Reply:
ah, i see what you mean. ultimately the issue lies in the fact that the `KeepWorkingBadge` you need for this context isn't actually a bubble badge (in that it's not attached to a bubble), as James called out. but i don't think it's worth digging into that right now.

your refactor looks good. one NIT is that imo `ProgressBubble` shouldn't need to import two components to render something so small and static, so you could clean this up a bit by exporting that snippet above as `AssessmentBubbleBadge`, but i'm fine with it as-is since you want to ship this today.